cmd/upgrade/cluster/cmd.go

```diff
@@ -128,24 +139,120 @@ func run(cmd *cobra.Command, _ []string) {
 		os.Exit(1)
 	}
 
-	scheduledUpgrade, upgradeState, err := r.OCMClient.GetScheduledUpgrade(cluster.ID())
+	if args.controlPlane && !isHypershift {
+		r.Reporter.Errorf("The '--control-plane' option is only supported for Hosted Control Planes")
+		os.Exit(1)
+	}
+
+	if !args.controlPlane && isHypershift {
+		r.Reporter.Errorf("The '--control-plane' option is currently mandatory for Hosted Control Planes")
+		os.Exit(1)
+	}
+
+	checkExistingScheduledUpgrade(r, cluster, clusterKey)
+
+	availableUpgrades, version := buildVersion(r, cmd, cluster, args.version)
+	err = r.OCMClient.CheckUpgradeClusterVersion(availableUpgrades, version, cluster)
 	if err != nil {
-		r.Reporter.Errorf("Failed to get scheduled upgrades for cluster '%s': %v", clusterKey, err)
+		r.Reporter.Errorf("%v", err)
 		os.Exit(1)
 	}
-	if scheduledUpgrade != nil {
-		r.Reporter.Warnf("There is already a %s upgrade to version %s on %s",
-			upgradeState.Value(),
-			scheduledUpgrade.Version(),
-			scheduledUpgrade.NextRun().Format("2006-01-02 15:04 MST"),
-		)
+
+	if scheduleDate == "" || scheduleTime == "" {
+		interactive.Enable()
+	}
+
+	if isSTS && mode == "" {
+		mode = setMode(r, cmd)
+	}
+
+	// if cluster is sts validate roles are compatible with upgrade version
+	if isSTS {
+		checkSTSRolesCompatibility(r, cluster, mode, version, clusterKey)
+	}
+
+	version, err = ocm.CheckAndParseVersion(availableUpgrades, version)
+	if err != nil {
+		r.Reporter.Errorf("Error parsing version to upgrade to")
+		os.Exit(1)
+	}
+	if !confirm.Confirm("upgrade cluster to version '%s'", version) {
 		os.Exit(0)
 	}
 
-	version := args.version
-	scheduleDate := args.scheduleDate
-	scheduleTime := args.scheduleTime
+	if isHypershift {
+		createUpgradePolicyHypershift(r, cmd, clusterKey, cluster, version, scheduleDate, scheduleTime)
+	} else {
+		createUpgradePolicyClassic(r, cmd, clusterKey, cluster, version, scheduleDate, scheduleTime)
+	}
 
+	// Drain grace period config
+	clusterSpec := buildNodeDrainGracePeriod(r, cmd, cluster)
+	err = r.OCMClient.UpdateCluster(cluster.ID(), r.Creator, clusterSpec)
+	if err != nil {
+		r.Reporter.Errorf("Failed to update cluster '%s': %v", clusterKey, err)
+		os.Exit(1)
+	}
+
+	r.Reporter.Infof("Upgrade successfully scheduled for cluster '%s'", clusterKey)
+}
+
+func createUpgradePolicyHypershift(r *rosa.Runtime, cmd *cobra.Command, clusterKey string,
+	cluster *cmv1.Cluster, version string, scheduleDate string, scheduleTime string) {
+	upgradePolicyBuilder := cmv1.NewControlPlaneUpgradePolicy().ScheduleType("manual").
+		UpgradeType("ControlPlane").Version(version)
+	// TODO: remove forceNow when we support scheduling
+	nextRun := buildUpgradeSchedule(r, cmd, scheduleDate, scheduleTime, true)
+	upgradePolicyBuilder = upgradePolicyBuilder.NextRun(nextRun)
+	upgradePolicy, err := upgradePolicyBuilder.Build()
+	if err != nil {
+		r.Reporter.Errorf("Failed to schedule upgrade for cluster '%s': %v", clusterKey, err)
```

Comment:
NP
We can return an error from the function, and exit the program from the main function.

Reply:
thanks, changed accordingly